File: day17/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

type (
	Point struct {
		x, y, z, w int
	}
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	active := map[Point]struct{}{}
	active2 := map[Point]struct{}{}

	scanner := bufio.NewScanner(file)
	for i := 0; scanner.Scan(); i++ {
		for j, c := range []byte(scanner.Text()) {
			if c == '#' {
				active[Point{j, i, 0, 0}] = struct{}{}
				active2[Point{j, i, 0, 0}] = struct{}{}
			}
		}
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	for i := 0; i < 6; i++ {
		active = step(active, false)
	}
	fmt.Println("Part 1:", len(active))
	for i := 0; i < 6; i++ {
		active2 = step(active2, true)
	}
	fmt.Println("Part 2:", len(active2))
}

func step(active map[Point]struct{}, dim4 bool) map[Point]struct{} {
	next := map[Point]struct{}{}
	nbc := map[Point]int{}
	for p := range active {
		var nb []Point
		if dim4 {
			nb = neighborsW(p)
		} else {
			nb = neighbors(p)
		}
		c := countNeighbors(nb, active)
		if c == 2 || c == 3 {
			next[p] = struct{}{}
		}
		for _, n := range nb {
			if _, ok := nbc[n]; !ok {
				nbc[n] = 0
			}
			nbc[n]++
		}
	}
	for p, c := range nbc {
		if c == 3 {
			next[p] = struct{}{}
		}
	}
	return next
}

func countNeighbors(points []Point, active map[Point]struct{}) int {
	count := 0
	for _, p := range points {
		if _, ok := active[p]; ok {
			count++
		}
	}
	return count
}

func neighbors(p Point) []Point {
	n := make([]Point, 0, 26)
	for i := p.x - 1; i <= p.x+1; i++ {
		for j := p.y - 1; j <= p.y+1; j++ {
			for k := p.z - 1; k <= p.z+1; k++ {
				x := Point{i, j, k, 0}
				if x != p {
					n = append(n, x)
				}
			}
		}
	}
	return n
}

func neighborsW(p Point) []Point {
	n := make([]Point, 0, 26)
	for i := p.x - 1; i <= p.x+1; i++ {
		for j := p.y - 1; j <= p.y+1; j++ {
			for k := p.z - 1; k <= p.z+1; k++ {
				for l := p.w - 1; l <= p.w+1; l++ {
					x := Point{i, j, k, l}
					if x == p {
						continue
					}
					n = append(n, x)
				}
			}
		}
	}
	return n
}

File: day17/src/main.rs
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BErr = Box<dyn std::error::Error>;

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
struct Point(i32, i32, i32, i32);

fn main() -> Result<(), BErr> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut active = HashSet::new();
    for (i, line) in reader.lines().enumerate() {
        let line = line?;
        for (j, c) in line.chars().enumerate() {
            if c == '#' {
                active.insert(Point(j as i32, i as i32, 0, 0));
            }
        }
    }
    let mut active2 = active.clone();

    for _ in 0..6 {
        active = step(&active, false);
    }
    println!("Part 1: {}", active.len());
    for _ in 0..6 {
        active2 = step(&active2, true);
    }
    println!("Part 2: {}", active2.len());
    Ok(())
}

fn step(active: &HashSet<Point>, dim4: bool) -> HashSet<Point> {
    let mut next = HashSet::new();
    let mut nbc = HashMap::new();
    for &p in active {
        let nb = if dim4 { neighbors_w(p) } else { neighbors(p) };
        let c = count_neighbors(&nb, active);
        if c == 2 || c == 3 {
            next.insert(p);
        }
        for n in nb {
            let v = nbc.entry(n).or_insert(0);
            *v += 1;
        }
    }
    for (p, c) in nbc {
        if c == 3 {
            next.insert(p);
        }
    }
    next
}

fn count_neighbors(points: &[Point], active: &HashSet<Point>) -> usize {
    let mut count = 0;
    for p in points {
        if active.contains(p) {
            count += 1;
        }
    }
    count
}

fn neighbors(p: Point) -> Vec<Point> {
    let mut points = Vec::with_capacity(26);
    for i in p.0 - 1..=p.0 + 1 {
        for j in p.1 - 1..=p.1 + 1 {
            for k in p.2 - 1..=p.2 + 1 {
                let x = Point(i, j, k, 0);
                if x != p {
                    points.push(x);
                }
            }
        }
    }
    points
}

fn neighbors_w(p: Point) -> Vec<Point> {
    let mut points = Vec::with_capacity(26);
    for i in p.0 - 1..=p.0 + 1 {
        for j in p.1 - 1..=p.1 + 1 {
            for k in p.2 - 1..=p.2 + 1 {
                for l in p.3 - 1..=p.3 + 1 {
                    let x = Point(i, j, k, l);
                    if x != p {
                        points.push(x);
                    }
                }
            }
        }
    }
    points
}
